valhalla/midgard/gridded_data.h

```diff
@@ -100,14 +116,339 @@ template <class coord_t> class GriddedData : public Tiles<coord_t> {
+      }
+    }
+
+    // we need something to hold each iso-line
+    contours_t contours(intervals.size(), std::list<feature_t>{feature_t{}});
+
+    // and something to find them quickly
+    using contour_lookup_t = std::unordered_map<PointLL, typename feature_t::iterator>;
+    std::vector<contour_lookup_t> lookups(intervals.size());
+    // TODO: preallocate the lookups for each interval
+
+    // For each metric we tracked
+    for (const auto& metric : metrics) {
+      size_t metric_index = std::get<0>(*metric.first);
+
+      // For each cell, skipping the outer rim since its out of bounds
+      for (int row = 1; row < this->nrows_ - 1; ++row) {
+        for (int col = 1; col < this->ncolumns_ - 1; ++col) {
+          int tileid = this->TileId(col, row);
+          auto cell1 = data_[tileid][metric_index];
+          auto cell2 = data_[tileid + this->ncolumns_][metric_index];     // TileId(col,   row+1)];
+          auto cell3 = data_[tileid + 1][metric_index];                   // TileId(col+1, row)];
+          auto cell4 = data_[tileid + this->ncolumns_ + 1][metric_index]; // TileId(col+1, row+1)];
+          auto dmin = std::min(std::min(cell1, cell2), std::min(cell3, cell4));
+          auto dmax = std::max(std::max(cell1, cell2), std::max(cell3, cell4));
+
+          // Continue if outside the range of contour values for this metric_index
+          if (dmax < std::get<1>(*std::prev(metric.second)) || dmin > std::get<1>(*metric.first)) {
+            continue;
+          }
+
+          // For each requested contour value
+          for (size_t i = 0; i < intervals.size(); ++i) {
+            // some setup to process this contour
+            auto& lookup = lookups[i];
+            auto& contour = contours[i];
+            auto contour_value = std::get<1>(intervals[i]);
+
+            // we skip this contour if its interested in a different metric_index or its value
+            // would not intersect this cell
+            if (std::get<0>(intervals[i]) != metric_index || contour_value < dmin ||
+                contour_value > dmax) {
+              continue;
+            }
+
+            for (int m = 4; m > 0; m--) {
+              int newtileid = tileid + tile_inc[m - 1];
+              // Make sure the tile corner value is not set to the max_value
+              // (messes up the intersect method). Set a value slightly above
+              // the contour (e.g. 1 minute higher).
+              // TODO - the value 1 is a bit of a hack.
+              float nd = data_[newtileid][metric_index];
+              s[m] = nd < max_value_[metric_index] ? nd - contour_value : 1.0f;
+              tile_corners[m] = this->Base(newtileid);
+              sh[m] = (s[m] > 0.0f) - (s[m] < 0.0f); // pos = 1, neg = -1, 0 = 0
+            }
+            s[0] = 0.25 * (s[1] + s[2] + s[3] + s[4]);
+            tile_corners[0] = this->Center(tileid);
+            sh[0] = (s[0] > 0.0f) - (s[0] < 0.0f); // pos = 1, neg = -1, 0 = 0
```

Comment:
previously this bit of code had 3 branches not including the loop condition. i rolled 0th element out of the loop to remove one and i removed the other 2 branches by using math to find the sign of the number